### PR TITLE
feat: add support for remaining fields in Kubernetes SD

### LINKED ISSUE #6087: Support other fields available in KubernetesSDConfigs
**What is missing?**

We didn't cover all fields in initial PR for KubernetesSDConfigs https://github.com/prometheus-operator/prometheus-operator/pull/5675, this request is to support remaining  fields in https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kubernetes_sd_config 

**Why do we need it?**

To complete the feature

**Environment**

* Prometheus Operator version:

    `Insert image tag or Git SHA here`
    
    

**Anything else we need to know?**:

### PR DESCRIPTION
**Pull request by slashpai**
Fixes #6087

## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
feat: add support for remaining fields in Kubernetes SD in ScrapeConfig
```

### CODE REVIEW

**Comment by slashpai** [comment]
@simonpasquier This should be good for review :)

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -2499,11 +2499,79 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+			configs[i] = append(configs[i], yaml.MapItem{
+				Key:   "role",
+				Value: strings.ToLower(string(config.Role)),
+			})
+
+			configs[i] = cg.addBasicAuthToYaml(configs[i], assetStoreKey, store, config.BasicAuth)
+			configs[i] = cg.addSafeAuthorizationToYaml(configs[i], fmt.Sprintf("scrapeconfig/auth/%s/%s/kubernetessdconfig/%d", sc.GetNamespace(), sc.GetName(), i), store, config.Authorization)
+			configs[i] = cg.addOAuth2ToYaml(configs[i], config.Oauth2, store.OAuth2Assets, assetStoreKey)
+
+			if config.ProxyConfig != nil {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2520]: there's a lot of repetition with existing code. We need a utility function that translates a ProxyConfig field into Prometheus YAML config.

File: pkg/prometheus/promcfg.go [hunk 2]
```diff
@@ -2499,11 +2499,79 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+					Value: config.EnableHttp2,
+				})
+			}
+
+			if config.TLSConfig != nil {
+				configs[i] = addSafeTLStoYaml(configs[i], sc.GetNamespace(), *config.TLSConfig)
+			}
+
+			if config.Namespaces != nil {
+				configs[i] = append(configs[i], yaml.MapItem{
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2571]: I'd rather iterate over every field than assuming an identical serialization between the CRD and Prometheus configuration.

File: pkg/prometheus/promcfg.go [hunk 3]
```diff
@@ -2530,6 +2598,17 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 					Value: selectors,
 				})
 			}
+
+			if config.AttachMetadata != nil {
+				switch config.Role {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2603]:
```suggestion
				switch strings.ToLower(config.Role) {
```

File: pkg/prometheus/promcfg.go [hunk 4]
```diff
@@ -2530,6 +2598,17 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+				case "pod", "Pod":
+					configs[i] = cg.WithMinimumVersion("2.35.0").AppendMapItem(configs[i], "attach_metadata", config.AttachMetadata)
+				case "endpoints", "Endpoints", "endpointslice", "EndpointSlice":
+					configs[i] = cg.WithMinimumVersion("2.37.0").AppendMapItem(configs[i], "attach_metadata", config.AttachMetadata)
+				default:
+					level.Warn(cg.logger).Log("msg", fmt.Sprintf("ignoring attachMetadata not supported by Prometheus for role: %s", config.Role))
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2609]: should we reject the resource in the first place?

Comment (slashpai) [reply pkg/prometheus/promcfg.go:2609]: I think its ok to warn than rejecting the resource. We do similar for other version specific newer fields.

File: pkg/prometheus/promcfg.go (hunk 1 above)
Comment (slashpai) [reply pkg/prometheus/promcfg.go:2520]: sure, I can do that in follow up PR since ProxyConfig is used in a few SDConfigs as well

**Comment by slashpai** [comment]
@simonpasquier addressed comments. could you please review again?

File: pkg/prometheus/promcfg.go (hunk 4 above)
Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:2618]: I'm on the fence too. Prometheus wouldn't fail either if you configure node's metadata with a service role so a warning is probably fine.
For clarity in the CRD description, it would be better to duplicate the struct from pkg/api/monitoring/v1 and state in the v1alpha1 comment that it's only valid for Pod, Endpoint and Endpointslice roles.

File: pkg/prometheus/resource_selector.go [hunk 5]
```diff
@@ -766,8 +766,38 @@ func (rs *ResourceSelector) SelectScrapeConfigs(ctx context.Context, listFn List
...
+			return fmt.Errorf("[%d]: %w", i, err)
+		}
+
+		if config.ProxyConfig != nil {
+			if err := validateProxyConfig(ctx, config.ProxyConfig, rs.store, sc.GetNamespace()); err != nil {
+				return fmt.Errorf("[%d]: %w", i, err)
+			}
+		}
+
+		if config.APIServer != nil && config.Namespaces != nil {
```
Comment (simonpasquier) [inline pkg/prometheus/resource_selector.go:795]:
```suggestion
		if config.APIServer != nil && len(config.Namespaces) > 0 {
```

Comment (slashpai) [reply pkg/prometheus/resource_selector.go:795]: config.Namespaces isn't array so we wouldn't be able to use `len()`

**Comment by slashpai** [comment]
@simonpasquier addressed comments

File: pkg/prometheus/promcfg.go [hunk 6]
```diff
@@ -2498,11 +2498,89 @@ func (cg *ConfigGenerator) generateScrapeConfig(
...
+
+			if config.TLSConfig != nil {
+				configs[i] = addSafeTLStoYaml(configs[i], sc.GetNamespace(), *config.TLSConfig)
+			}
+
+			if config.Namespaces != nil {
+				namespaces := []yaml.MapItem{
+					{
+						Key:   "own_namespace",
+						Value: config.Namespaces.IncludeOwnNamespace,
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2573]: we need to test IncludeOwnNamespace against nil

Comment (slashpai) [reply pkg/prometheus/promcfg.go:2573]: I added two more tests for this

Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:2573]: I mean that when IncludeOwnNamespace it shouldn't appear in the generated config.

**Review by simonpasquier — APPROVED** [review]
